theano/compile/function_module.py

```diff
@@ -488,16 +489,195 @@ def __setitem__(self, item, value):
+            """
+            assert sv_ori.type == sv_rpl.type, (
+                "Type of given SharedVariable conflicts with origianl one",
+                "Type of given SharedVariable:", sv_rpl.type,
+                "Type of original SharedVariable:", sv_ori.type)
+
+        maker = self.maker
+
+        # Copy Ins and their storage.
+        # so that they have different storage as their value
+        ins = [copy.copy(input) for input in maker.inputs]
+
+        # Delete update output in fgraph and updates In instances if needed
+        if delete_updates:
+            # The first len(maker.outputs) variabels are original variables.
+            # The rest are the updates.
+            out_vars = maker.fgraph.outputs[:len(maker.outputs)]
+        else:
+            out_vars = maker.fgraph.outputs
+
+        # Init new fgraph using copied variables and get memo
+        # memo: a dict that map old variables to new variables
+        memo = graph.clone_get_equiv(maker.fgraph.inputs, out_vars)
+        fg_cpy = gof.fg.FunctionGraph([memo[i] for i in maker.fgraph.inputs],
+                                      [memo[o] for o in out_vars],
+                                      clone=False)
+
+        # Re initialize Outs and swap update and variable in Ins
+        # By doing this, we can pass FunctionMaker._check_unused_inputs()
+        outs = list(map(SymbolicOutput, fg_cpy.outputs[:len(maker.outputs)]))
+        for out_ori, out_cpy in zip(maker.outputs, outs):
+            out_cpy.borrow = out_ori.borrow
+
+        # swap SharedVariable
+        if swap is not None:
+            swap_svs_ori = swap.keys()
+            exist_svs = [i.variable for i in maker.inputs]
+
+            # Check if given ShareVariables exist
+            for sv in swap_svs_ori:
+                if sv not in exist_svs:
+                    raise ValueError("SharedVariable: %s not found" %
+                                     (sv.name))
+
+            # Swap SharedVariable in fgraph and In instances
+            for index, (i, in_v) in enumerate(zip(ins, fg_cpy.inputs)):
+                # Variables in maker.inputs are defined by user, therefore we
+                # use them to make comparision and do the mapping.
+                # Otherwise we don't touch them.
+                var = maker.inputs[index].variable
+
+                if var in swap_svs_ori:
+                    swap_sv = swap[var]
+                    checkSV(i.variable, swap_sv)
+
+                    # swap variable and value of In instances
+                    i.variable = swap_sv
+                    i.value = swap_sv.container
+
+                    # In the fgraph we use the cloned SharedVariable
+                    swap_sv = swap_sv.clone()
+
+                    # Swap SharedVariable in fgraph
+                    # if inputs was replaced, change self.inputs
+                    fg_cpy.inputs[index] = swap_sv
+                    fg_cpy.replace(in_v, swap_sv, reason="Swap SV")
```

Comment:
This seems to make it possible to replace an ordinary input by a shared variable through the copy.  I wonder if that is intended.